qiskit_experiments/whole_backend/partition.py

```diff
@@ -0,0 +1,222 @@
+"""
+Functions to partition qubits and edges into groups, such that qubits/edges
+in a group are distance from each other
+"""
+
+import retworkx as rx
+
+
+def distance_graph(graph, distance):
+    """
+    The vertices of the distance graph are the same as
+    in the original graph. Two vertices are connected by an edge if the distance between them
+    in the original graph is smaller than `distance`.
+    """
+    distance_matrix = rx.distance_matrix(graph)
+    dist_graph = rx.PyGraph()
+    indexes = graph.node_indices()
+
+    for graph_vertex1 in indexes:
+        dist_graph.add_node(graph[graph_vertex1])
+        for graph_vertex2 in range(graph_vertex1):
+            if distance_matrix[graph_vertex1, graph_vertex2] < distance:
+                dist_graph.add_edge(graph_vertex1, graph_vertex2, None)
```

Comment:
This code assumes that `graph` has ordered node indices starting from 0 and that the node data values are the same as the indices. That is the case for the graphs generated in this file as long as it is the case for `graph = rx.PyGraph(); graph.add_nodes_from(list(range(N)))` (which is not documented behavior but seems fairly safe). It might be good to document those assumptions or to do more handling of indices to remove the assumptions. For example, if `graph.remove_node` had been called before `graph` was passed to this function, there might be a missing index number and then the graph indices and the indices of the distance matrix will not match.

I think this addresses the edge case I am describing (but there might be a cleaner way):

```suggestion
    distance_matrix = rx.distance_matrix(graph)
    dist_graph = rx.PyGraph()
    indexes = graph.node_indices()
    distance_indexes = []

    for idx1, graph_vertex1 in enumerate(indexes):
        dist_idx = dist_graph.add_node(graph[graph_vertex1])
        distance_indexes.append(dist_idx)
        for idx2 in range(idx1):
            if distance_matrix[idx1, idx2] < distance:
                dist_graph.add_edge(disance_indexes[idx1], distance_indexes[idx2], None)
```

I was trying to partition the device in a different way starting from this code and ran into something like this, so I thought I would mention it.

Reply:
This was fixed in the internal repo